images/virtualization-artifact/pkg/controller/vi/internal/source/object_ref.go

```diff
@@ -32,35 +36,204 @@ import (
 )
 
+const objectRefDataSource = "objectref"
+
 type ObjectRefDataSource struct {
 	statService     Stat
 	importerService Importer
 	dvcrSettings    *dvcr.Settings
 	client          client.Client
+	diskService     *service.DiskService
+
+	viObjectRefOnPvc *ObjectRefDataVirtualImageOnPVC
 }
 
 func NewObjectRefDataSource(
 	statService Stat,
 	importerService Importer,
 	dvcrSettings *dvcr.Settings,
 	client client.Client,
+	diskService *service.DiskService,
 ) *ObjectRefDataSource {
 	return &ObjectRefDataSource{
-		statService:     statService,
-		importerService: importerService,
-		dvcrSettings:    dvcrSettings,
-		client:          client,
+		statService:      statService,
+		importerService:  importerService,
+		dvcrSettings:     dvcrSettings,
+		client:           client,
+		diskService:      diskService,
+		viObjectRefOnPvc: NewObjectRefDataVirtualImageOnPVC(statService, importerService, dvcrSettings, client, diskService),
 	}
 }
 
-func (ds ObjectRefDataSource) Sync(ctx context.Context, vi *virtv2.VirtualImage) (bool, error) {
+func (ds ObjectRefDataSource) StoreToPVC(ctx context.Context, vi *virtv2.VirtualImage) (bool, error) {
+	log, ctx := logger.GetDataSourceContext(ctx, objectRefDataSource)
+
+	if vi.Spec.DataSource.ObjectRef.Kind == virtv2.VirtualImageKind {
+		viKey := types.NamespacedName{Name: vi.Spec.DataSource.ObjectRef.Name, Namespace: vi.Namespace}
+		viObjetcRef, err := helper.FetchObject(ctx, viKey, ds.client, &virtv2.VirtualImage{})
+		if err != nil {
+			return false, fmt.Errorf("unable to get VI %s: %w", viKey, err)
+		}
+
+		if viObjetcRef.Spec.Storage == virtv2.StorageKubernetes {
+			return ds.viObjectRefOnPvc.StoreToPVC(ctx, vi, viObjetcRef)
+		}
+	}
+
+	condition, _ := service.GetCondition(vdcondition.ReadyType, vi.Status.Conditions)
+	defer func() { service.SetCondition(condition, &vi.Status.Conditions) }()
+
+	supgen := supplements.NewGenerator(cc.VIShortName, vi.Name, vi.Namespace, vi.UID)
+	dv, err := ds.diskService.GetDataVolume(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+	pvc, err := ds.diskService.GetPersistentVolumeClaim(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+
+	switch {
+	case isDiskProvisioningFinished(condition):
+		log.Info("Disk provisioning finished: clean up")
+
+		setPhaseConditionForFinishedImage(pvc, &condition, &vi.Status.Phase, supgen)
+
+		// Protect Ready Disk and underlying PVC.
+		err = ds.diskService.Protect(ctx, vi, nil, pvc)
+		if err != nil {
+			return false, err
+		}
+
+		err = ds.diskService.Unprotect(ctx, dv)
+		if err != nil {
+			return false, err
+		}
+
+		return CleanUpSupplements(ctx, vi, ds)
+	case cc.AnyTerminating(dv, pvc):
+		log.Info("Waiting for supplements to be terminated")
+	case dv == nil:
+		log.Info("Start import to PVC")
+
+		var dvcrDataSource controller.DVCRDataSource
+		dvcrDataSource, err = controller.NewDVCRDataSourcesForVMI(ctx, vi.Spec.DataSource, vi, ds.client)
```

Comment:
dvcrDataSource.GetTarget returns "" for vi with Kubernetes storage

Reply:
I didn't understand why

Reply:
Add handle this case

Reply:
Add.